src/Compiler/Microsoft.CodeAnalysis.Razor/src/BindTagHelperDescriptorProvider.cs

```diff
@@ -262,72 +272,78 @@ private List<ElementBindData> GetElementBindData(Compilation compilation)
             }
         }
 
-        var results = new List<ElementBindData>();
+        using var results = new PooledArrayBuilder<ElementBindData>();
+
+        (string Assembly, string Type, string Namespace)? displayNames;
 
-        for (var i = 0; i < types.Count; i++)
+        foreach (var type in types)
         {
-            var type = types[i];
-            var attributes = type.GetAttributes();
+            // Set this to null, since we have a new type.
+            displayNames = null;
 
             // Not handling duplicates here for now since we're the primary ones extending this.
             // If we see users adding to the set of 'bind' constructs we will want to add deduplication
             // and potentially diagnostics.
-            for (var j = 0; j < attributes.Length; j++)
+            foreach (var attribute in type.GetAttributes())
             {
-                var attribute = attributes[j];
+                // Be careful to only compute the display names once for each type.
+                displayNames ??= (type.ContainingAssembly.Name, type.ToDisplayString(), type.ContainingNamespace.ToDisplayString());
```

Comment:
>displayNames ??= (type.ContainingAssembly.Name, type.ToDisplayString(), type.ContainingNamespace.ToDisplayString()); [](http://example.com/codeflow?start=16&length=116)

consider moving this inside the if conditions so it doesn't have to allocate for types which don't ever add to results

Reply:
Yeah, I can make this better. Thanks!